commons-math/src/main/java/org/apache/commons/math4/genetics/AbstractListChromosome.java

```diff
@@ -0,0 +1,106 @@
+     * @param representation inner representation of the chromosome
+     * @throws InvalidRepresentationException iff the <code>representation</code> can not represent a valid chromosome
+     */
+    public AbstractListChromosome(final List<T> representation) throws InvalidRepresentationException {
+        this(representation, true);
+    }
+
+    /**
+     * Constructor, copying the input representation.
+     * @param representation inner representation of the chromosome
+     * @throws InvalidRepresentationException iff the <code>representation</code> can not represent a valid chromosome
+     */
+    public AbstractListChromosome(final T[] representation) throws InvalidRepresentationException {
+        this(Arrays.asList(representation));
+    }
+
+    /**
+     * Constructor.
+     * @param representation inner representation of the chromosome
+     * @param copyList if {@code true}, the representation will be copied, otherwise it will be referenced.
+     * @since 3.3
+     */
+    public AbstractListChromosome(final List<T> representation, final boolean copyList) {
+        checkValidity(representation);
+        this.representation =
+                Collections.unmodifiableList(copyList ? new ArrayList<>(representation) : representation);
+    }
+
+    /**
+     * Asserts that <code>representation</code> can represent a valid chromosome.
+     *
+     * @param chromosomeRepresentation representation of the chromosome
+     * @throws InvalidRepresentationException iff the <code>representation</code> can not represent a valid chromosome
+     */
+    protected abstract void checkValidity(List<T> chromosomeRepresentation) throws InvalidRepresentationException;
+
+    /**
+     * Returns the (immutable) inner representation of the chromosome.
+     * @return the representation of the chromosome
+     */
+    protected List<T> getRepresentation() {
+        return representation;
+    }
+
+    /**
+     * Returns the length of the chromosome.
+     * @return the length of the chromosome
+     */
+    public int getLength() {
+        return getRepresentation().size();
+    }
+
+    /**
+     * Creates a new instance of the same class as <code>this</code> is, with a given <code>arrayRepresentation</code>.
+     * This is needed in crossover and mutation operators, where we need a new instance of the same class, but with
+     * different array representation.
+     * <p>
+     * Usually, this method just calls a constructor of the class.
+     *
+     * @param chromosomeRepresentation the inner array representation of the new chromosome.
+     * @return new instance extended from FixedLengthChromosome with the given arrayRepresentation
+     */
+    public abstract AbstractListChromosome<T> newFixedLengthChromosome(final List<T> chromosomeRepresentation);
+
+    /** {@inheritDoc} */
+    @Override
+    public String toString() {
+        return String.format("(f=%s %s)", getFitness(), getRepresentation());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format double: &#39;getFitness&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 104. Consider formatting this data with &#39;%f&#39; instead.